// @forestadmin/datasource-customizer/dist/decorators/search/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const collection_search_context_1 = __importDefault(require("./collection-search-context"));
const normalize_name_1 = __importDefault(require("./normalize-name"));
const parse_query_1 = require("./parse-query");
class SearchCollectionDecorator extends datasource_toolkit_1.CollectionDecorator {
    constructor() {
        super(...arguments);
        this.replacer = null;
    }
    replaceSearch(replacer) {
        this.replacer = replacer;
    }
    refineSchema(subSchema) {
        return { ...subSchema, searchable: true };
    }
    async refineFilter(caller, filter) {
        // Search string is not significant
        if (!filter?.search?.trim().length) {
            return filter?.override({ search: null });
        }
        // Implement search ourselves
        if (this.replacer || !this.childCollection.schema.searchable) {
            const ctx = new collection_search_context_1.default(this, caller, this.generatePlainSearchFilter.bind(this, caller));
            let tree;
            if (this.replacer) {
                const plainTree = await this.replacer(filter.search, filter.searchExtended, ctx);
                tree = datasource_toolkit_1.ConditionTreeFactory.fromPlainObject(plainTree);
            }
            else {
                tree = this.generateSearchFilter(caller, filter.search, {
                    extended: filter.searchExtended,
                });
            }
            // Note that if no fields are searchable with the provided searchString, the conditions
            // array might be empty, which will create a condition returning zero records
            // (this is the desired behavior).
            return filter.override({
                conditionTree: datasource_toolkit_1.ConditionTreeFactory.intersect(filter.conditionTree, tree),
                search: null,
            });
        }
        // Let subcollection deal with the search
        return filter;
    }
    generateSearchFilter(caller, searchText, options) {
        const parsedQuery = (0, parse_query_1.parseQuery)(searchText);
        const specifiedFields = options?.onlyFields ? [] : (0, parse_query_1.extractSpecifiedFields)(parsedQuery);
        const defaultFields = options?.onlyFields
            ? []
            : this.getFields(this.childCollection, Boolean(options?.extended));
        const searchableFields = new Map([
            ...defaultFields,
            ...[...specifiedFields, ...(options?.onlyFields ?? []), ...(options?.includeFields ?? [])]
                .map(name => this.lenientGetSchema(name))
                .filter(Boolean)
                .map(schema => [schema.field, schema.schema]),
        ]
            .filter(Boolean)
            .filter(([field]) => !options?.excludeFields?.includes(field)));
        const conditionTree = (0, parse_query_1.generateConditionTree)(caller, parsedQuery, [...searchableFields]);
        if (!conditionTree && searchText.trim().length) {
            return datasource_toolkit_1.ConditionTreeFactory.MatchNone;
        }
        return conditionTree;
    }
    generatePlainSearchFilter(caller, searchText, options) {
        const conditionTree = this.generateSearchFilter(caller, searchText, options);
        return conditionTree?.toPlainObject();
    }
    getFields(collection, extended) {
        const fields = [];
        for (const [name, field] of Object.entries(collection.schema.fields)) {
            if (field.type === 'Column')
                fields.push([name, field]);
            if (extended && (field.type === 'ManyToOne' || field.type === 'OneToOne')) {
                const related = collection.dataSource.getCollection(field.foreignCollection);
                for (const [subName, subField] of Object.entries(related.schema.fields)) {
                    if (subField.type === 'Column')
                        fields.push([`${name}:${subName}`, subField]);
                }
            }
        }
        return fields;
    }
    lenientGetSchema(path) {
        const [prefix, suffix] = path.split(/:(.*)/);
        const fuzzyPrefix = (0, normalize_name_1.default)(prefix);
        for (const [field, schema] of Object.entries(this.schema.fields)) {
            const fuzzyFieldName = (0, normalize_name_1.default)(field);
            if (fuzzyPrefix === fuzzyFieldName) {
                if (!suffix && schema.type === 'Column') {
                    return { field, schema };
                }
                if (suffix &&
                    (schema.type === 'OneToMany' || schema.type === 'ManyToOne' || schema.type === 'OneToOne')) {
                    const related = this.dataSource.getCollection(schema.foreignCollection);
                    const fuzzy = related.lenientGetSchema(suffix);
                    if (fuzzy)
                        return { field: `${field}:${fuzzy.field}`, schema: fuzzy.schema };
                }
            }
        }
        return null;
    }
}
exports.default = SearchCollectionDecorator;
