// @forestadmin/datasource-toolkit/dist/src/base-datasource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class BaseDataSource {
    constructor() {
        this._collections = {};
    }
    get collections() {
        return Object.values(this._collections);
    }
    get schema() {
        return { charts: [] };
    }
    getCollection(name) {
        const collection = this._collections[name];
        if (collection === undefined) {
            throw new Error(`Collection '${name}' not found. List of available collections: ${Object.keys(this._collections)
                .sort()
                .join(', ')}`);
        }
        return collection;
    }
    addCollection(collection) {
        if (this._collections[collection.name] !== undefined) {
            throw new Error(`Collection '${collection.name}' already defined in datasource`);
        }
        this._collections[collection.name] = collection;
    }
    renderChart(caller, name) {
        throw new Error(`No chart named '${name}' exists on this datasource.`);
    }
}
exports.default = BaseDataSource;
