// @forestadmin/datasource-toolkit/dist/src/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntrospectionFormatError = exports.ForbiddenError = exports.UnprocessableError = exports.ValidationError = exports.BusinessError = void 0;
// eslint-disable-next-line max-classes-per-file
class BusinessError extends Error {
    constructor(message, data, name) {
        super(message);
        this.name = name ?? this.constructor.name;
        this.data = data;
    }
    /**
     * We cannot rely on `instanceof` because there can be some mismatch between
     * packages versions as dependencies of different packages.
     * So this function is a workaround to check if an error is of a specific type.
     */
    static isOfType(error, ErrorConstructor) {
        return error.name === ErrorConstructor.name;
    }
}
exports.BusinessError = BusinessError;
class ValidationError extends BusinessError {
}
exports.ValidationError = ValidationError;
class UnprocessableError extends BusinessError {
}
exports.UnprocessableError = UnprocessableError;
class ForbiddenError extends BusinessError {
}
exports.ForbiddenError = ForbiddenError;
class IntrospectionFormatError extends BusinessError {
    constructor(sourcePackageName) {
        const message = `This version of introspection is newer than this package version. ` +
            `Please update ${sourcePackageName}`;
        super(message);
    }
    /** @deprecated use name instead */
    get type() {
        return this.name;
    }
}
exports.IntrospectionFormatError = IntrospectionFormatError;
