// @forestadmin/forestadmin-client/dist/permissions/forest-http-api.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const auth_1 = __importDefault(require("../auth"));
const server_1 = __importDefault(require("../utils/server"));
class ForestHttpApi {
    async getEnvironmentPermissions(options) {
        return server_1.default.query(options, 'get', '/liana/v4/permissions/environment');
    }
    async getUsers(options) {
        return server_1.default.query(options, 'get', '/liana/v4/permissions/users');
    }
    async getRenderingPermissions(renderingId, options) {
        return server_1.default.query(options, 'get', `/liana/v4/permissions/renderings/${renderingId}`);
    }
    async getModelCustomizations(options) {
        return server_1.default.query(options, 'get', '/liana/model-customizations');
    }
    makeAuthService(options) {
        return new auth_1.default(options);
    }
}
exports.default = ForestHttpApi;
