// @forestadmin/datasource-toolkit/dist/src/validation/rules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAP_ALLOWED_TYPES_FOR_OPERATOR_CONDITION_TREE = exports.MAP_ALLOWED_TYPES_FOR_COLUMN_TYPE = exports.MAP_ALLOWED_OPERATORS_FOR_COLUMN_TYPE = void 0;
const BASE_OPERATORS = ['Blank', 'Equal', 'Missing', 'NotEqual', 'Present'];
const ARRAY_OPERATORS = ['In', 'NotIn', 'IncludesAll', 'IncludesNone'];
const BASE_DATEONLY_OPERATORS = [
    'Today',
    'Yesterday',
    'PreviousXDaysToDate',
    'PreviousWeek',
    'PreviousWeekToDate',
    'PreviousMonth',
    'PreviousMonthToDate',
    'PreviousQuarter',
    'PreviousQuarterToDate',
    'PreviousYear',
    'PreviousYearToDate',
    'Past',
    'Future',
    'PreviousXDays',
    'Before',
    'After',
];
exports.MAP_ALLOWED_OPERATORS_FOR_COLUMN_TYPE = Object.freeze({
    String: [
        ...BASE_OPERATORS,
        ...ARRAY_OPERATORS,
        'Contains',
        'NotContains',
        'EndsWith',
        'StartsWith',
        'LongerThan',
        'ShorterThan',
        'Like',
        'ILike',
        'IContains',
        'NotIContains',
        'IEndsWith',
        'IStartsWith',
    ],
    Number: [...BASE_OPERATORS, ...ARRAY_OPERATORS, 'GreaterThan', 'LessThan'],
    Dateonly: [...BASE_OPERATORS, ...BASE_DATEONLY_OPERATORS],
    Date: [...BASE_OPERATORS, ...BASE_DATEONLY_OPERATORS, 'BeforeXHoursAgo', 'AfterXHoursAgo'],
    Time: [...BASE_OPERATORS, 'LessThan', 'GreaterThan'],
    Timeonly: [...BASE_OPERATORS, 'LessThan', 'GreaterThan'],
    Binary: [...BASE_OPERATORS, ...ARRAY_OPERATORS],
    Enum: [...BASE_OPERATORS, ...ARRAY_OPERATORS],
    Json: [...BASE_OPERATORS, ...ARRAY_OPERATORS],
    Boolean: [...BASE_OPERATORS, ...ARRAY_OPERATORS],
    Point: BASE_OPERATORS,
    Uuid: [...BASE_OPERATORS, ...ARRAY_OPERATORS],
});
exports.MAP_ALLOWED_TYPES_FOR_COLUMN_TYPE = Object.freeze({
    String: ['String', null],
    Binary: ['Binary', null],
    Number: ['Number', null],
    Boolean: ['Boolean', null],
    Enum: ['Enum', null],
    Date: ['Date', null],
    Dateonly: ['Dateonly', null],
    Json: ['Json', null],
    Point: ['Point', null],
    Time: ['Time', null],
    Timeonly: ['Timeonly', null],
    Uuid: ['Uuid', null],
});
const defaultsOperators = Object.keys(exports.MAP_ALLOWED_OPERATORS_FOR_COLUMN_TYPE).reduce((mapMemo, type) => {
    const allowedOperators = exports.MAP_ALLOWED_OPERATORS_FOR_COLUMN_TYPE[type];
    allowedOperators.forEach(operator => {
        if (mapMemo[operator]) {
            mapMemo[operator].push(type);
        }
        else {
            mapMemo[operator] = [type];
        }
    });
    return mapMemo;
}, {});
const NO_TYPES_ALLOWED = [null];
exports.MAP_ALLOWED_TYPES_FOR_OPERATOR_CONDITION_TREE = Object.freeze({
    ...defaultsOperators,
    In: [...defaultsOperators.In, null],
    NotIn: [...defaultsOperators.NotIn, null],
    IncludesAll: [...defaultsOperators.IncludesAll, null],
    IncludesNone: [...defaultsOperators.IncludesNone, null],
    Blank: NO_TYPES_ALLOWED,
    Missing: NO_TYPES_ALLOWED,
    Present: NO_TYPES_ALLOWED,
    Yesterday: NO_TYPES_ALLOWED,
    Today: NO_TYPES_ALLOWED,
    PreviousQuarter: NO_TYPES_ALLOWED,
    PreviousYear: NO_TYPES_ALLOWED,
    PreviousMonth: NO_TYPES_ALLOWED,
    PreviousWeek: NO_TYPES_ALLOWED,
    Past: NO_TYPES_ALLOWED,
    Future: NO_TYPES_ALLOWED,
    PreviousWeekToDate: NO_TYPES_ALLOWED,
    PreviousMonthToDate: NO_TYPES_ALLOWED,
    PreviousQuarterToDate: NO_TYPES_ALLOWED,
    PreviousYearToDate: NO_TYPES_ALLOWED,
    PreviousXDaysToDate: ['Number'],
    PreviousXDays: ['Number'],
    BeforeXHoursAgo: ['Number'],
    AfterXHoursAgo: ['Number'],
    LongerThan: ['Number'],
    ShorterThan: ['Number'],
});
