// @forestadmin/datasource-customizer/dist/decorators/search/custom-parser/condition-tree-query-walker.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const index_1 = __importDefault(require("../filter-builder/index"));
const QueryListener_1 = __importDefault(require("../generated-parser/QueryListener"));
const normalize_name_1 = __importDefault(require("../normalize-name"));
class ConditionTreeQueryWalker extends QueryListener_1.default {
    get conditionTree() {
        if (this.parentStack.length !== 1 && this.parentStack[0].length !== 1) {
            throw new Error('Invalid condition tree');
        }
        return this.parentStack[0][0];
    }
    constructor(caller, fields) {
        super();
        this.caller = caller;
        this.fields = fields;
        this.parentStack = [];
        this.currentField = null;
        this.isNegated = false;
        this.enterQuery = () => {
            this.parentStack.push([]);
        };
        this.exitQuery = () => {
            const rules = this.parentStack.pop();
            if (!rules) {
                throw new Error('Empty query');
            }
            if (rules.length === 1) {
                this.parentStack.push(rules);
            }
            else {
                this.parentStack.push([datasource_toolkit_1.ConditionTreeFactory.intersect(...rules)]);
            }
        };
        this.exitQuoted = (ctx) => {
            const current = this.parentStack[this.parentStack.length - 1];
            current.push(this.buildDefaultCondition(ctx.getText().slice(1, -1), this.isNegated));
        };
        this.enterNegated = () => {
            this.parentStack.push([]);
            this.isNegated = true;
        };
        this.exitNegated = (ctx) => {
            const text = ctx.getText();
            const rules = this.parentStack.pop();
            if (!rules)
                return;
            let result;
            if (!Number.isNaN(Number(text)) && rules.length === 1) {
                result = this.buildDefaultCondition(text, false);
            }
            else {
                result = datasource_toolkit_1.ConditionTreeFactory.intersect(...rules.filter(Boolean));
            }
            const parentRules = this.parentStack[this.parentStack.length - 1];
            if (parentRules) {
                parentRules.push(result);
            }
            else {
                // We should at least have an array for the root query
                throw new Error('Empty stack');
            }
            this.isNegated = false;
        };
        this.exitWord = (ctx) => {
            const current = this.parentStack[this.parentStack.length - 1];
            current.push(this.buildDefaultCondition(ctx.getText(), this.isNegated));
        };
        this.enterPropertyMatching = (ctx) => {
            this.currentField = ctx.getChild(0).getText().replace(/\./g, ':');
        };
        this.exitPropertyMatching = () => {
            this.currentField = null;
        };
        this.enterOr = () => {
            this.parentStack.push([]);
        };
        this.exitOr = () => {
            const rules = this.parentStack.pop();
            if (!rules.length)
                return;
            const parentRules = this.parentStack[this.parentStack.length - 1];
            parentRules.push(datasource_toolkit_1.ConditionTreeFactory.union(...rules));
        };
        this.enterAnd = () => {
            this.parentStack.push([]);
        };
        this.exitAnd = () => {
            const rules = this.parentStack.pop();
            if (!rules.length)
                return;
            const parentRules = this.parentStack[this.parentStack.length - 1];
            parentRules.push(datasource_toolkit_1.ConditionTreeFactory.intersect(...rules));
        };
    }
    generateDefaultFilter(searchQuery) {
        return this.buildDefaultCondition(searchQuery, this.isNegated);
    }
    buildDefaultCondition(searchString, isNegated) {
        const targetedFields = this.currentField &&
            this.fields.filter(([field]) => (0, normalize_name_1.default)(field) === (0, normalize_name_1.default)(this.currentField));
        let rules = [];
        if (!targetedFields?.length) {
            rules = this.fields.map(([field, schema]) => (0, index_1.default)(this.caller, field, schema, 
            // If targetFields is empty, it means that the query is not targeting a specific field
            // OR that the field is not found in the schema. If it's the case, we are re-constructing
            // the original query by adding the field name in front of the search string.
            this.currentField ? `${this.currentField}:${searchString}` : searchString, isNegated));
        }
        else {
            rules = targetedFields.map(([field, schema]) => (0, index_1.default)(this.caller, field, schema, searchString, isNegated));
        }
        if (!rules.some(Boolean))
            return null;
        return isNegated
            ? datasource_toolkit_1.ConditionTreeFactory.intersect(...rules)
            : datasource_toolkit_1.ConditionTreeFactory.union(...rules);
    }
}
exports.default = ConditionTreeQueryWalker;
