// @forestadmin/datasource-customizer/dist/decorators/search/filter-builder/build-date-field-filter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable no-continue */
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const luxon_1 = require("luxon");
const build_default_condition_1 = __importDefault(require("./utils/build-default-condition"));
function isYear(str) {
    return (/^\d{4}$/.test(str) && Number(str) >= 1800 && Number(str) <= new Date().getFullYear() + 100);
}
function isYearMonth(str) {
    return /^(\d{4})-(\d{1,2})$/.test(str) && !Number.isNaN(Date.parse(`${str}-01`));
}
function isPlainDate(str) {
    return /^\d{4}-\d{2}-\d{2}$/.test(str) && !Number.isNaN(Date.parse(str));
}
function isValidDate(str) {
    return isYear(str) || isYearMonth(str) || isPlainDate(str);
}
function getPeriodStart(string) {
    if (isYear(string))
        return `${string}-01-01`;
    if (isYearMonth(string))
        return `${string}-01`;
    return string;
}
function pad(month) {
    if (month < 10) {
        return `0${month}`;
    }
    return `${month}`;
}
function getAfterPeriodEnd(string) {
    if (isYear(string))
        return `${Number(string) + 1}-01-01`;
    if (isYearMonth(string)) {
        const [year, month] = string.split('-').map(Number);
        const endDate = new Date(year, month, 1);
        return `${endDate.getFullYear()}-${pad(endDate.getMonth() + 1)}-01`;
    }
    const date = new Date(string);
    date.setDate(date.getDate() + 1);
    return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}
function getTranslatedDateInTimezone(value, opts) {
    if (opts.columnType !== 'Date')
        return value;
    return luxon_1.DateTime.fromISO(value, { zone: opts.timezone }).toISO();
}
const supportedOperators = [
    [
        '>',
        [
            ['After', getAfterPeriodEnd],
            ['Equal', getAfterPeriodEnd],
        ],
        [
            ['Before', getPeriodStart],
            ['Equal', getPeriodStart],
            ['Missing', () => undefined],
        ],
    ],
    [
        '>=',
        [
            ['After', getPeriodStart],
            ['Equal', getPeriodStart],
        ],
        [
            ['Before', getPeriodStart],
            ['Missing', () => undefined],
        ],
    ],
    [
        '≥',
        [
            ['After', getPeriodStart],
            ['Equal', getPeriodStart],
        ],
        [
            ['Before', getPeriodStart],
            ['Missing', () => undefined],
        ],
    ],
    [
        '<',
        [['Before', getPeriodStart]],
        [
            ['After', getPeriodStart],
            ['Equal', getPeriodStart],
            ['Missing', () => undefined],
        ],
    ],
    [
        '<=',
        [['Before', getAfterPeriodEnd]],
        [
            ['After', getAfterPeriodEnd],
            ['Equal', getAfterPeriodEnd],
            ['Missing', () => undefined],
        ],
    ],
    [
        '≤',
        [['Before', getAfterPeriodEnd]],
        [
            ['After', getAfterPeriodEnd],
            ['Equal', getAfterPeriodEnd],
            ['Missing', () => undefined],
        ],
    ],
];
function buildDateFieldFilter({ field, filterOperators, searchString, isNegated, columnType, timezone, }) {
    if (isValidDate(searchString)) {
        const start = getTranslatedDateInTimezone(getPeriodStart(searchString), {
            columnType,
            timezone,
        });
        const afterEnd = getTranslatedDateInTimezone(getAfterPeriodEnd(searchString), {
            columnType,
            timezone,
        });
        if (!isNegated &&
            filterOperators.has('Equal') &&
            filterOperators.has('Before') &&
            filterOperators.has('After')) {
            return datasource_toolkit_1.ConditionTreeFactory.intersect(datasource_toolkit_1.ConditionTreeFactory.union(new datasource_toolkit_1.ConditionTreeLeaf(field, 'Equal', start), new datasource_toolkit_1.ConditionTreeLeaf(field, 'After', start)), new datasource_toolkit_1.ConditionTreeLeaf(field, 'Before', afterEnd));
        }
        if (isNegated &&
            filterOperators.has('Before') &&
            filterOperators.has('After') &&
            filterOperators.has('Equal')) {
            if (filterOperators.has('Missing')) {
                return datasource_toolkit_1.ConditionTreeFactory.union(new datasource_toolkit_1.ConditionTreeLeaf(field, 'Before', start), new datasource_toolkit_1.ConditionTreeLeaf(field, 'After', afterEnd), new datasource_toolkit_1.ConditionTreeLeaf(field, 'Equal', afterEnd), new datasource_toolkit_1.ConditionTreeLeaf(field, 'Missing'));
            }
            return datasource_toolkit_1.ConditionTreeFactory.union(new datasource_toolkit_1.ConditionTreeLeaf(field, 'Before', start), new datasource_toolkit_1.ConditionTreeLeaf(field, 'After', afterEnd), new datasource_toolkit_1.ConditionTreeLeaf(field, 'Equal', afterEnd));
        }
        return (0, build_default_condition_1.default)(isNegated);
    }
    for (const [operatorPrefix, positiveOperations, negativeOperations] of supportedOperators) {
        if (!searchString.startsWith(operatorPrefix))
            continue;
        if (!isValidDate(searchString.slice(operatorPrefix.length)))
            continue;
        const value = searchString.slice(operatorPrefix.length);
        if (!isValidDate(value))
            continue;
        const operations = isNegated ? negativeOperations : positiveOperations;
        // If Missing is not supported, we try to build a condition tree anyway
        if (!operations
            .filter(op => op[0] !== 'Missing')
            .every(operation => filterOperators.has(operation[0]))) {
            continue;
        }
        return datasource_toolkit_1.ConditionTreeFactory.union(...operations
            .filter(op => filterOperators.has(op[0]))
            .map(([operator, getDate]) => new datasource_toolkit_1.ConditionTreeLeaf(field, operator, getTranslatedDateInTimezone(getDate(value), { timezone, columnType }))));
    }
    return (0, build_default_condition_1.default)(isNegated);
}
exports.default = buildDateFieldFilter;
