// @forestadmin/datasource-customizer/dist/decorators/sort-emulate/collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
class SortEmulate extends datasource_toolkit_1.CollectionDecorator {
    constructor() {
        super(...arguments);
        this.sorts = new Map();
        this.disabledSorts = new Set();
    }
    emulateFieldSorting(name) {
        this.replaceOrEmulateFieldSorting(name, null);
    }
    /**
     * Disable sorting on this field. This only prevents the end-user to sort on this field.
     * It will still be possible to sort on this field in the customizations code.
     * @param name name of the field
     * @deprecated this method will be removed soon.
     */
    disableFieldSorting(name) {
        datasource_toolkit_1.FieldValidator.validate(this, name);
        this.disabledSorts.add(name);
        this.markSchemaAsDirty();
    }
    replaceFieldSorting(name, equivalentSort) {
        if (!equivalentSort) {
            throw new Error('A new sorting method should be provided to replace field sorting');
        }
        this.replaceOrEmulateFieldSorting(name, equivalentSort);
    }
    replaceOrEmulateFieldSorting(name, equivalentSort) {
        datasource_toolkit_1.FieldValidator.validate(this, name);
        this.sorts.set(name, equivalentSort ? new datasource_toolkit_1.Sort(...equivalentSort) : null);
        this.markSchemaAsDirty();
    }
    async list(caller, filter, projection) {
        const childFilter = filter.override({
            sort: filter.sort?.replaceClauses(clause => this.rewritePlainSortClause(clause)),
        });
        if (!childFilter.sort?.some(({ field }) => this.isEmulated(field))) {
            return this.childCollection.list(caller, childFilter, projection);
        }
        // Fetch the whole collection, but only with the fields we need to sort
        let referenceRecords;
        referenceRecords = await this.childCollection.list(caller, childFilter.override({ sort: null, page: null }), childFilter.sort.projection.withPks(this));
        referenceRecords = childFilter.sort.apply(referenceRecords);
        if (childFilter.page)
            referenceRecords = childFilter.page.apply(referenceRecords);
        // We now have the information we need to sort by the field
        const newFilter = new datasource_toolkit_1.Filter({
            conditionTree: datasource_toolkit_1.ConditionTreeFactory.matchRecords(this.schema, referenceRecords),
        });
        let records;
        records = await this.childCollection.list(caller, newFilter, projection.withPks(this));
        records = this.sortRecords(referenceRecords, records);
        records = projection.apply(records);
        return records;
    }
    refineSchema(childSchema) {
        const fields = {};
        for (const [name, schema] of Object.entries(childSchema.fields)) {
            if (schema.type === 'Column') {
                let sortable = schema.isSortable;
                if (this.disabledSorts.has(name)) {
                    // disableFieldSorting
                    sortable = false;
                }
                else if (this.sorts.has(name)) {
                    // replaceFieldSorting
                    sortable = true;
                }
                fields[name] = { ...schema, isSortable: sortable };
            }
            else {
                fields[name] = schema;
            }
        }
        return { ...childSchema, fields };
    }
    sortRecords(referenceRecords, records) {
        const positionById = {};
        const sorted = new Array(records.length);
        for (const [index, record] of referenceRecords.entries()) {
            positionById[datasource_toolkit_1.RecordUtils.getPrimaryKey(this.schema, record).join('|')] = index;
        }
        for (const record of records) {
            const id = datasource_toolkit_1.RecordUtils.getPrimaryKey(this.schema, record).join('|');
            sorted[positionById[id]] = record;
        }
        return sorted;
    }
    rewritePlainSortClause(clause) {
        // Order by is targeting a field on another collection => recurse.
        if (clause.field.includes(':')) {
            const [prefix] = clause.field.split(':');
            const schema = this.schema.fields[prefix];
            const association = this.dataSource.getCollection(schema.foreignCollection);
            return new datasource_toolkit_1.Sort(clause)
                .unnest()
                .replaceClauses(subClause => association.rewritePlainSortClause(subClause))
                .nest(prefix);
        }
        // Field that we own: recursively replace using equivalent sort
        let equivalentSort = this.sorts.get(clause.field);
        if (equivalentSort) {
            if (!clause.ascending)
                equivalentSort = equivalentSort.inverse();
            return equivalentSort.replaceClauses(subClause => this.rewritePlainSortClause(subClause));
        }
        return new datasource_toolkit_1.Sort(clause);
    }
    isEmulated(path) {
        const index = path.indexOf(':');
        if (index === -1)
            return this.sorts.has(path);
        const { foreignCollection } = this.schema.fields[path.substring(0, index)];
        const association = this.dataSource.getCollection(foreignCollection);
        return association.isEmulated(path.substring(index + 1));
    }
}
exports.default = SortEmulate;
