// @forestadmin/datasource-customizer/dist/decorators/override/collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const context_1 = require("./context");
class OverrideCollectionDecorator extends datasource_toolkit_1.CollectionDecorator {
    async create(caller, data) {
        if (this.createHandler) {
            const context = new context_1.CreateOverrideCustomizationContext(this.childCollection, caller, data);
            return this.createHandler(context);
        }
        return super.create(caller, data);
    }
    addCreateHandler(handler) {
        this.createHandler = handler;
    }
    async update(caller, filter, patch) {
        if (this.updateHandler) {
            const context = new context_1.UpdateOverrideCustomizationContext(this.childCollection, caller, filter, patch);
            return this.updateHandler(context);
        }
        return super.update(caller, filter, patch);
    }
    addUpdateHandler(handler) {
        this.updateHandler = handler;
    }
    async delete(caller, filter) {
        if (this.deleteHandler) {
            const context = new context_1.DeleteOverrideCustomizationContext(this.childCollection, caller, filter);
            return this.deleteHandler(context);
        }
        return super.delete(caller, filter);
    }
    addDeleteHandler(handler) {
        this.deleteHandler = handler;
    }
}
exports.default = OverrideCollectionDecorator;
