// @forestadmin/datasource-customizer/dist/decorators/override/context.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteOverrideCustomizationContext = exports.UpdateOverrideCustomizationContext = exports.CreateOverrideCustomizationContext = void 0;
const collection_context_1 = __importDefault(require("../../context/collection-context"));
class CreateOverrideCustomizationContext extends collection_context_1.default {
    constructor(collection, caller, data) {
        super(collection, caller);
        this.data = data;
    }
}
exports.CreateOverrideCustomizationContext = CreateOverrideCustomizationContext;
class UpdateOverrideCustomizationContext extends collection_context_1.default {
    constructor(collection, caller, filter, patch) {
        super(collection, caller);
        this.filter = filter;
        this.patch = patch;
    }
}
exports.UpdateOverrideCustomizationContext = UpdateOverrideCustomizationContext;
class DeleteOverrideCustomizationContext extends collection_context_1.default {
    constructor(collection, caller, filter) {
        super(collection, caller);
        this.filter = filter;
    }
}
exports.DeleteOverrideCustomizationContext = DeleteOverrideCustomizationContext;
